### PR TITLE
feat: ordered mention suggestions

### PR DESCRIPTION
**Pull request by gijoe0295**
### Details

Implement mention suggestions with order of relevancy when in a group chat or a chat connected to a workspace:

1. People within the chat (members of the Report)
2. People within the Workspace
3. Other Contacts

### Fixed Issues

$ https://github.com/Expensify/App/issues/42009
PROPOSAL: https://github.com/Expensify/App/issues/42009#issuecomment-2105413569

### Tests

1. Type `@` in a group chat with several participants
2. Verify the order of suggestions: group members >> other contacts, both in alphabetical order
3. Type in a letter
4. Verify the list is filtered to the individuals containing that letter with the same hierarchy preserved as described above
---------
1. Type `@` in a Workspace room, where only a fraction of the workspace members are in the room
2. Verify the order of suggestions: room members >> workspace members >> other contacts, all in alphabetical order
3. Type in a letter
4. Verify the list is filtered to the individuals containing that letter with the same hierarchy preserved as described above

- [x] Verify that no errors appear in the JS console

### Offline tests

NA

### QA Steps

Same as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

image
image
image
image

</details>

<details>
<summary>Android: mWeb Chrome</summary>

image
image

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/153004152/da1cca36-c925-43d1-8524-c00f514797d5

Simulator Screenshot - iPhone 15 Pro Max - 2024-05-24 at 02 16 36

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - iPhone 15 Pro Max - 2024-05-24 at 02 15 58

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

#### Group chat
Screenshot 2024-05-24 at 02 08 42

Screenshot 2024-05-24 at 02 11 12

#### Workspace room

Screenshot 2024-05-24 at 01 59 12
Screenshot 2024-05-24 at 01 59 24

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by alitoshmatov** [comment]
@gijoe0295 Can you check this in groups, in my case suggestion is not showing report members first

Screenshot 2024-05-26 at 10 30 07 PM
Screenshot 2024-05-26 at 10 30 12 PM

**Comment by alitoshmatov** [comment]
@gijoe0295 Can't add to an existing group

**Comment by alitoshmatov** [comment]
Looks like this is happening only with old groups. I couldn't reproduce the issue when creating a new croup even choosing the same members

File: src/pages/home/report/ReportActionCompose/SuggestionMention.tsx [hunk 1]
```diff
@@ -58,6 +69,36 @@ function SuggestionMention(
...
+            }
+            return 2;
+        },
+        [currentReport],
+    );
+    const weightedPersonalDetails: PersonalDetailsList | SuggestionPersonalDetailsList = useMemo(() => {
+        const policyEmployeeAccountIDs = getPolicyEmployeeAccountIDs(policyID);
+        if (!ReportUtils.isGroupChat(currentReport) && !ReportUtils.doesReportBelongToWorkspace(currentReport, policyEmployeeAccountIDs, policyID)) {
+            return personalDetails;
+        }
```
[the remark this answers is not in the document]
Comment (alitoshmatov) [reply src/pages/home/report/ReportActionCompose/SuggestionMention.tsx:91]: Sorry, I confused the cases in the condition, that's why my conclusion was wrong

I see that for a report to be a group chat it should have `chatType=group` but it is empty for this case. Maybe it is a thing with older groups. I checked this with multiple old groups in two accounts and the same true for all

Below is `currentReport,  policyEmployeeAccountIDs, policyID`
```js
{
    "reportID": "1164088907436698",
    "reportName": "Chat Report",
    "type": "chat",
    "chatType": "",
    "ownerAccountID": 0,
    "managerID": 0,
    "policyID": "_FAKE_",
    "participants": {
        "14579716": {
            "hidden": false
        },
        "14582230": {
            "hidden": false
        },
        "14640187": {
            "hidden": false
        },
        "14640366": {
            "hidden": false
        },
        "14640421": {
            "hidden": false
        }
    },
    "participantAccountIDs": [
        14582230,
        14640187,
        14640366,
        14640421
    ],
    "visibleChatMemberAccountIDs": [
        14582230,
        14640187,
        14640366,
        14640421
    ],
    "isPinned": false,
    "lastReadTime": "2024-04-12 15:49:29.672",
    "lastReadSequenceNumber": 0,
    "lastVisibleActionCreated": "2024-04-12 15:49:29.672",
    "lastVisibleActionLastModified": "2024-04-12 15:49:29.672",
    "lastMessageText": "@alitoshmatov2001+test@gmail.com  Hello world, new @alitoshmatov2001+android@gm…",
    "lastActionType": "ADDCOMMENT",
    "lastActorAccountID": "14579716",
    "notificationPreference": "always",
    "stateNum": 0,
    "statusNum": 0,
    "oldPolicyName": "",
    "isOwnPolicyExpenseChat": false,
    "lastMessageHtml": "<mention-user>@alitoshmatov2001+test@gmail.com</mention-user>  Hello world, new <mention-user>@alitoshmatov2001+android@gmail.com</mention-user>",
    "hasOutstandingChildRequest": false,
    "writeCapability": "all",
    "description": "",
    "total": 0,
    "unheldTotal": 0,
    "currency": "USD",
    "isWaitingOnBankAccount": false,
    "nonReimbursableTotal": 0,
    "isCancelledIOU": false,
    "permissions": [
        "read",
        "write"
    ],
    "submitterPayPalMeAddress": "",
    "welcomeMessage": "",
    "errorFields": {}
}
```

Screenshot 2024-06-05 at 3 19 25 AM

[the remark this answers is not in the document]
Comment (alitoshmatov) [reply src/pages/home/report/ReportActionCompose/SuggestionMention.tsx:91]: I think we should proceed with this PR

[the remark this answers is not in the document]
Comment (yuwenmemon) [reply src/pages/home/report/ReportActionCompose/SuggestionMention.tsx:91]: @gijoe0295 Yeah maybe let's raise in expensify-bugs: https://expensify.enterprise.slack.com/archives/C049HHMV9SM

**Comment by alitoshmatov** [comment]
I will complete the checklist

**Review by alitoshmatov — APPROVED** [review]
LGTM!

**Review by yuwenmemon — APPROVED** [review]

File: src/pages/home/report/ReportActionCompose/SuggestionMention.tsx [hunk 2]
```diff
@@ -58,6 +69,36 @@ function SuggestionMention(
...
+        return lodashMapValues(personalDetails, (detail) =>
+            detail
+                ? {
+                      ...detail,
+                      weight: getPersonalDetailsWeight(detail, policyEmployeeAccountIDs),
```
Comment (yuwenmemon) [inline src/pages/home/report/ReportActionCompose/SuggestionMention.tsx:96]: NAB: A more apt name could be maybe something like `orderWeight`

**Review by yuwenmemon — APPROVED** [review]

**Pull request merged**